docs/guides/developer/docs/api/workflow-api.md

```diff
@@ -0,0 +1,585 @@
+
+# Workflow instances
+
+### GET /api/workflows
+
+Returns a list of workflow instances.
+
+The following query string parameters are supported to filter, sort and pagingate the returned list:
+
+Query String Parameter | Type                        | Description
+:----------------------|:----------------------------|:-----------
+`filter`               | [`string`](types.md#basic)  | A comma-separated list of filters to limit the results with (see [Filtering](usage.md#filtering)). See the below table for the list of available filters
+`sort`                 | [`string`](types.md#basic)  | A comma-separated list of sort criteria (see [Sorting](usage.md#sorting)).  See the below table for the list of available sort criteria
+`limit`                | [`integer`](types.md#basic) | The maximum number of results to return (see [Pagination](usage.md#pagination))
+`offset`               | [`integer`](types.md#basic) | The index of the first result to return (see [Pagination](usage.md#pagination))
+
+The following filters are available:
+
+Filter Name                      | Description
+:--------------------------------|:-----------
+`state`                          | Workflow instances that are in this state. Can be included multiple times (inclusive or).
+`state_not`                      | Workflow instances that are not in this state. Can be included multiple times (inclusive or).
+`current_operation`              | Workflow instances that are currently executing this operation. Can be included multiple times (inclusive or).
+`current_operation_not`          | Workflow instances that are currently not executing this operation. Can be included multiple times (inclusive or).
+`workflow_definition_identifier` | Workflow instances that use this workflow definition
+`event_identifier`               | Workflow instances where the identifier of the processed event matches
+`event_title`                    | Workflow instances where the title of the processed event matches
+`event_created`                  | Workflow instances where the processed event was created between two dates
+`event_creator`                  | Workflow instances where the creator of the processed event matches
+`event_contributor`              | Workflow instances where the contributor of the processed event matches
+`event_language`                 | Workflow instances where the language of the processed event matches
+`event_license`                  | Workflow instances where the license of the processed event matches
+`event_subject`                  | Workflow instances where the subject of the processed event matches
+`series_identifier`              | Workflow instances where the identifier of series of the processed event matches
+`series_title`                   | Workflow instances where the title of series of the processed event matches
+`textFilter`                     | Workflow instances where any part of the Workflow definition's metadata fields match this value
```

Comment:
Hi @karendolan 

I don't have a strong opinion about which exact fields workflows can be filtered or sorted by, but I still think that  mixing up different kinds of objects (workflows vs. events vs. series) is undesirable from an architectural point of view.
Think of our database that will be never able to deliver answers to such filters/sort options (unless we duplicate the metadata in the workflow database table which we hopefully don't) and the need to have another data source ("search index") just to accumulate data of different services to be able to filter/sort across service boundaries. Or why do we have a search index for workflows at all?

I'm talking about the following fields...

- event_title
- event_created
- event_creator
- event_contributor
- event_language
- event_license
- event_subject
- series_identifier
- series_title

... which are not fields of a workflow instance to my understanding  (The event identifier in means of media package identifier is, as the workflow instance is bound to a media package).

I'm fine if this added to the External API if somebody actually needs this, but could we at least try to keep it up to an minimum?

I see that things like event_title, series_identifier and series_title can be useful. But do we really need all the other filters & sort options like event_language, for example?

Can you please have a look at this?

What do you think? Include all those fields? Can some be "safely" omitted?

You might guess that having all those copies of the same metadata values at various locations across the entire system and (trying to) keep them in sync is not exactly what I love most about Opencast... ;-)

Reply:
Just a quick note: As noted in the `FIXME`, sorting by `event_identifier` currently leads to an exception. Solr complains that it `can not sort on undefined field: mediapackageid_sort`. I have created [MH-13066](https://opencast.jira.com/browse/MH-13066) to track the issue. Fixing this is outside the scope of this pull request and I also don't know what the difference between this and the other fields is.